assets/systemd-unit-upgrader/teleport-upgrade

```diff
@@ -0,0 +1,299 @@
+        log_debug "cannot check maintenance window (no schedule)."
+        return 1
+    fi
+
+    time_now="$(date +%s)"
+    
+    while read line; do
+        prefix="$(echo $line | awk '{print $1}')"
+        if [[ "$prefix" != "win" ]]; then
+            continue
+        fi
+
+        start_time="$(echo $line | awk '{print $2}')"
+        end_time="$(echo $line | awk '{print $3}')" 
+
+        if (( time_now > start_time && time_now < end_time )); then
+            log_debug "within maintenance window ${start_time} - ${end_time}."
+            return 0
+        fi
+    done <<< "$upgrade_schedule"
+
+    log_debug "not within maintenance window."
+    return 1
+}
+
+# schedule_appears_healthy checks if schedule was defined and contains at least
+# one upgrade window that ends in the future.
+schedule_appears_healthy() {
+    if [[ $upgrade_schedule == "none" ]]; then
+        log_debug "schedule appears unhealthy (does not exist)."
+        return 1
+    fi
+
+    time_now="$(date +%s)"
+    
+    while read line; do
+        end_time="$(echo $line | awk '{print $2}')" 
+
+        if (( time_now < end_time )); then
+            log_debug "schedule appears healthy (contains future window)"
+            return 0
+        fi
+    done <<< "$upgrade_schedule"
+
+    log_debug "schedule appears unhealthy (contains no future windows)."
+    return 1
+}
+
+
+# get_current_version gets the current version of the teleport binary. if this
+# operation fails, we assume that teleport is eligible for upgrade.
+get_current_version() {
+    echo "1.2.3"
+    return 0
+    if cv_output="$(teleport version --raw)"; then
+        echo "${cv_output}"
+        return 0
+    fi
+
+    return 1
+}
+
+# upgrade_endpoint_fetch loads the specified value from the upgrade endpoint. the only
+# currently supported values are 'version' and 'critical'.
+upgrade_endpoint_fetch() {
+    host_path="${upgrade_endpoint}/${1}" 
+    if is_insecure_mode; then
+        log_warn "fetching $host_path in insecure mode (not safe for production use)."
+        if if_output="$(curl -sSf "http://${host_path}")"; then
+            echo "$if_output"
+            return 0
+        else
+            log_debug "failed to GET $host_path (insecure mode)."
+            return 1
+        fi
+    fi
+
+    log_debug "fetching $host_path..."
+
+    if sf_output="$(curl --proto '=https' --tlsv1.2 -sSf "https://${host_path}")"; then
+        echo "$sf_output"
+        return 0
+    else
+        log_debug "failed to GET $host_path"
+        return 1
+    fi
+}
+
+# get_target_version loads the current value of the /version endpoint.
+get_target_version() {
+    if tv_output="$(upgrade_endpoint_fetch version)"; then
+        echo "${tv_output#v}"
+        return 0
+    fi
+    return 1
+}
+
+# is_critical checks if the /critical endpoint is available and set to 'yes'.
+is_critical() {
+    if ef_output="$(upgrade_endpoint_fetch critical)"; then
+        if [[ $ef_output == "yes" ]]; then
```

Comment:
You may want to trim whitespace (spaces, carriage returns, line feeds) and maybe do a case-insensitive comparison.